File: src/main/java/org/jcodec/codecs/h264/decode/FrameReader.java
package org.jcodec.codecs.h264.decode;

import static org.jcodec.codecs.h264.H264Utils.unescapeNAL;
import static org.jcodec.codecs.h264.io.model.NALUnitType.IDR_SLICE;
import static org.jcodec.codecs.h264.io.model.NALUnitType.NON_IDR_SLICE;
import static org.jcodec.codecs.h264.io.model.NALUnitType.PPS;
import static org.jcodec.codecs.h264.io.model.NALUnitType.SPS;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

import org.jcodec.codecs.common.biari.MDecoder;
import org.jcodec.codecs.h264.decode.aso.MapManager;
import org.jcodec.codecs.h264.decode.aso.Mapper;
import org.jcodec.codecs.h264.io.CABAC;
import org.jcodec.codecs.h264.io.CAVLC;
import org.jcodec.codecs.h264.io.model.NALUnit;
import org.jcodec.codecs.h264.io.model.PictureParameterSet;
import org.jcodec.codecs.h264.io.model.SeqParameterSet;
import org.jcodec.codecs.h264.io.model.SliceHeader;
import org.jcodec.common.IntObjectMap;
import org.jcodec.common.io.BitReader;
import org.jcodec.common.io.NIOUtils;
import org.jcodec.common.logging.Logger;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * MPEG 4 AVC ( H.264 ) Frame reader
 * 
 * Conforms to H.264 ( ISO/IEC 14496-10 ) specifications
 * 
 * @author The JCodec project
 * 
 */
public class FrameReader {
    private IntObjectMap<SeqParameterSet> sps;
    private IntObjectMap<PictureParameterSet> pps;

    public FrameReader() {
        this.sps = new IntObjectMap<SeqParameterSet>();
        this.pps = new IntObjectMap<PictureParameterSet>();
    }

    public List<SliceReader> readFrame(List<ByteBuffer> nalUnits) {
        List<SliceReader> result = new ArrayList<SliceReader>();

        for (ByteBuffer nalData : nalUnits) {
            NALUnit nalUnit = NALUnit.read(nalData);

            unescapeNAL(nalData);
            if (SPS == nalUnit.type) {
                SeqParameterSet _sps = SeqParameterSet.read(nalData);
                sps.put(_sps.seq_parameter_set_id, _sps);
            } else if (PPS == nalUnit.type) {
                PictureParameterSet _pps = PictureParameterSet.read(nalData);
                pps.put(_pps.pic_parameter_set_id, _pps);
            } else if (IDR_SLICE == nalUnit.type || NON_IDR_SLICE == nalUnit.type) {
                if (sps.size() == 0 || pps.size() == 0) {
                    Logger.warn("Skipping frame as no SPS/PPS have been seen so far...");
                    return null;
                }
                result.add(createSliceReader(nalData, nalUnit));
            }
        }

        return result;
    }

    private SliceReader createSliceReader(ByteBuffer segment, NALUnit nalUnit) {
        BitReader _in = BitReader.createBitReader(segment);
        SliceHeaderReader shr = new SliceHeaderReader();
        SliceHeader sh = shr.readPart1(_in);
        sh.pps = pps.get(sh.pic_parameter_set_id);
        sh.sps = sps.get(sh.pps.seq_parameter_set_id);
        shr.readPart2(sh, nalUnit, sh.sps, sh.pps, _in);

        Mapper mapper = new MapManager(sh.sps, sh.pps).getMapper(sh);

        CAVLC[] cavlc = new CAVLC[] { new CAVLC(sh.sps, sh.pps, 2, 2), new CAVLC(sh.sps, sh.pps, 1, 1),
                new CAVLC(sh.sps, sh.pps, 1, 1) };

        int mbWidth = sh.sps.pic_width_in_mbs_minus1 + 1;
        CABAC cabac = new CABAC(mbWidth);

        MDecoder mDecoder = null;
        if (sh.pps.entropy_coding_mode_flag) {
            _in.terminate();
            int[][] cm = new int[2][1024];
            int qp = sh.pps.pic_init_qp_minus26 + 26 + sh.slice_qp_delta;
            cabac.initModels(cm, sh.slice_type, sh.cabac_init_idc, qp);
            mDecoder = new MDecoder(segment, cm);
        }

        return new SliceReader(sh.pps, cabac, cavlc, mDecoder, _in, mapper, sh, nalUnit);
    }
    
    public void addSpsList(List<ByteBuffer> spsList) {
        for (ByteBuffer byteBuffer : spsList) {
            addSps(byteBuffer);
        }
    }

    public void addSps(ByteBuffer byteBuffer) {
        ByteBuffer clone = NIOUtils.clone(byteBuffer);
        unescapeNAL(clone);
        SeqParameterSet s = SeqParameterSet.read(clone);
        sps.put(s.seq_parameter_set_id, s);
    }

    public void addPpsList(List<ByteBuffer> ppsList) {
        for (ByteBuffer byteBuffer : ppsList) {
            addPps(byteBuffer);
        }
    }

    public void addPps(ByteBuffer byteBuffer) {
        ByteBuffer clone = NIOUtils.clone(byteBuffer);
        unescapeNAL(clone);
        PictureParameterSet p = PictureParameterSet.read(clone);
        pps.put(p.pic_parameter_set_id, p);
    }
}

File: src/main/java/org/jcodec/codecs/h264/decode/FrameReader.kt
package org.jcodec.codecs.h264.decode

import org.jcodec.codecs.common.biari.MDecoder
import org.jcodec.codecs.h264.H264Utils
import org.jcodec.codecs.h264.decode.SliceHeaderReader.readPart1
import org.jcodec.codecs.h264.decode.SliceHeaderReader.readPart2
import org.jcodec.codecs.h264.decode.aso.MapManager
import org.jcodec.codecs.h264.io.CABAC
import org.jcodec.codecs.h264.io.CAVLC
import org.jcodec.codecs.h264.io.model.NALUnit
import org.jcodec.codecs.h264.io.model.NALUnitType
import org.jcodec.codecs.h264.io.model.PictureParameterSet
import org.jcodec.codecs.h264.io.model.SeqParameterSet
import org.jcodec.common.IntObjectMap
import org.jcodec.common.io.BitReader
import org.jcodec.common.io.NIOUtils
import org.jcodec.common.logging.Logger
import java.nio.ByteBuffer
import java.util.*

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * MPEG 4 AVC ( H.264 ) Frame reader
 *
 * Conforms to H.264 ( ISO/IEC 14496-10 ) specifications
 *
 * @author The JCodec project
 */
class FrameReader {
    private val sps: IntObjectMap<SeqParameterSet>
    private val pps: IntObjectMap<PictureParameterSet>
    fun readFrame(nalUnits: List<ByteBuffer?>): List<SliceReader>? {
        val result: MutableList<SliceReader> = ArrayList()
        for (nalData in nalUnits) {
            val nalUnit = NALUnit.read(nalData!!)
            H264Utils.unescapeNAL(nalData)
            if (NALUnitType.SPS == nalUnit.type) {
                val _sps = SeqParameterSet.read(nalData)
                sps.put(_sps.seqParameterSetId, _sps)
            } else if (NALUnitType.PPS == nalUnit.type) {
                val _pps = PictureParameterSet.read(nalData)
                pps.put(_pps.picParameterSetId, _pps)
            } else if (NALUnitType.IDR_SLICE == nalUnit.type || NALUnitType.NON_IDR_SLICE == nalUnit.type) {
                if (sps.size() == 0 || pps.size() == 0) {
                    Logger.warn("Skipping frame as no SPS/PPS have been seen so far...")
                    return null
                }
                result.add(createSliceReader(nalData, nalUnit))
            }
        }
        return result
    }

    private fun createSliceReader(segment: ByteBuffer?, nalUnit: NALUnit): SliceReader {
        val _in = BitReader.createBitReader(segment)
        val sh = readPart1(_in)
        sh.pps = pps[sh.picParameterSetId]
        sh.sps = sps[sh.pps!!.seqParameterSetId]
        readPart2(sh, nalUnit, sh.sps!!, sh.pps!!, _in)
        val mapper = MapManager(sh.sps!!, sh.pps!!).getMapper(sh)
        val cavlc = arrayOf(CAVLC(sh.sps, sh.pps, 2, 2), CAVLC(sh.sps, sh.pps, 1, 1),
                CAVLC(sh.sps, sh.pps, 1, 1))
        val mbWidth = sh.sps!!.picWidthInMbsMinus1 + 1
        val cabac = CABAC(mbWidth)
        var mDecoder: MDecoder? = null
        if (sh.pps!!.isEntropyCodingModeFlag) {
            _in.terminate()
            val cm = Array(2) { IntArray(1024) }
            val qp = sh.pps!!.picInitQpMinus26 + 26 + sh.sliceQpDelta
            cabac.initModels(cm, sh.sliceType!!, sh.cabacInitIdc, qp)
            mDecoder = MDecoder(segment, cm)
        }
        return SliceReader(sh.pps, cabac, cavlc, mDecoder, _in, mapper, sh, nalUnit)
    }

    fun addSpsList(spsList: List<ByteBuffer?>) {
        for (byteBuffer in spsList) {
            addSps(byteBuffer)
        }
    }

    fun addSps(byteBuffer: ByteBuffer?) {
        val clone = NIOUtils.clone(byteBuffer)
        H264Utils.unescapeNAL(clone)
        val s = SeqParameterSet.read(clone)
        sps.put(s.seqParameterSetId, s)
    }

    fun addPpsList(ppsList: List<ByteBuffer?>) {
        for (byteBuffer in ppsList) {
            addPps(byteBuffer)
        }
    }

    fun addPps(byteBuffer: ByteBuffer?) {
        val clone = NIOUtils.clone(byteBuffer)
        H264Utils.unescapeNAL(clone)
        val p = PictureParameterSet.read(clone)
        pps.put(p.picParameterSetId, p)
    }

    init {
        sps = IntObjectMap()
        pps = IntObjectMap()
    }
}